File: ios/Mynah/UI/Screens/UsageScreen.swift
import SwiftUI
import UIKit

/// What is left to spend, what a top-up costs, and what this month has run to.
///
/// The money on this screen is deliberately two different things kept apart.
/// The headline is calls — the unit anybody actually buys and spends. Below it,
/// unchanged, is what those calls cost us to run, itemised per call: that is not
/// a bill, and the footnote says so. Showing it is a choice — most services hide
/// their margin — but this app asks people to trust it with a phone line, and a
/// cost breakdown they can check against their own call list is worth more than
/// the margin is worth concealing.
///
/// One route to pay, not three. Android sells packs through Play's billing
/// library and through Stripe's own sheet; here the payment page is the whole
/// shop. An in-app sheet needs Stripe's SDK, and a StoreKit purchase needs a
/// server endpoint that checks a receipt with Apple — neither exists yet, and a
/// button that collects money the server cannot verify is worse than a link that
/// works. The page also takes WeChat Pay and Alipay, and is the only route that
/// lets somebody else pay.
struct UsageScreen: View {
    @ObservedObject var model: CallsViewModel
    let onBack: () -> Void

    @State private var copied = false
    @State private var now = Date()

    private var monthStart: Date {
        let parts = Calendar.current.dateComponents([.year, .month], from: now)
        return Calendar.current.date(from: parts) ?? now
    }

    private var thisMonth: [Call] {
        let start = Int(monthStart.timeIntervalSince1970 * 1000)
        return model.calls.filter { $0.createdAt >= start }.sorted { $0.createdAt > $1.createdAt }
    }

    private var priced: [(call: Call, cost: Cost)] {
        thisMonth.compactMap { call in call.cost.map { (call, $0) } }
    }

    /// The currency is whatever Twilio rated in, so it is only known once at
    /// least one call has been rated.
    private var symbol: String {
        switch (priced.first?.cost.unit ?? "").uppercased() {
        case "GBP": "£"
        case "USD": "$"
        case "EUR": "€"
        default: ""
        }
    }

    /// Until then there is no total to show: a bare "0.00" reads as "this month
    /// was free" when it means "not priced yet".
    private var totalLabel: String {
        guard !priced.isEmpty else { return t("cost_pending_short") }
        let total = priced.reduce(0.0) { $0 + (Double($1.cost.price) ?? 0) }
        return "\(symbol)\(String(format: "%.2f", total))"
    }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(t("usage_title"), onBack: onBack)

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    balanceCard

                    // --- the shop ---
                    Spacer().frame(height: 16)
                    GroupLabel(t("usage_shop")).padding(.bottom, 8)

                    let packs = model.usage.inAppPacks
                    if packs.isEmpty, model.usage.packs.isEmpty, !model.usage.webPay {
                        Text(t("usage_shop_unavailable"))
                            .wise(Type.caption)
                            .foregroundStyle(Ink.mute)
                            .fixedSize(horizontal: false, vertical: true)
                    } else if !packs.isEmpty {
                        GroupedCard {
                            ForEach(packs) { pack in
                                PackRow(calls: pack.calls, price: pack.price, onBuy: openPaymentPage)
                            }
                        }
                    } else if !model.usage.packs.isEmpty {
                        // No price attached: those come from Play on Android and
                        // there is nothing here to ask. The page shows them.
                        GroupedCard {
                            ForEach(model.usage.packs) { pack in
                                PackRow(calls: pack.calls, price: nil, onBuy: openPaymentPage)
                            }
                        }
                    }

                    if model.usage.webPay {
                        Spacer().frame(height: 10)
                        GroupedCard { payRow }
                        Spacer().frame(height: 8)
                        Text(t(copied ? "usage_other_copied" : "usage_other_note"))
                            .wise(Type.tiny)
                            .foregroundStyle(copied ? Ink.positiveDeep : Ink.mute)
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.horizontal, 4)
                    }

                    // --- what the month cost to run ---
                    Spacer().frame(height: 22)
                    GroupLabel(t("usage_breakdown")).padding(.bottom, 8)

                    if thisMonth.isEmpty {
                        Text(t("usage_empty"))
                            .wise(Type.caption)
                            .foregroundStyle(Ink.mute)
                            .fixedSize(horizontal: false, vertical: true)
                    } else {
                        GroupedCard {
                            ForEach(thisMonth) { call in
                                UsageRow(call: call, symbol: symbol)
                            }
                            TotalRow(
                                seconds: priced.reduce(0) { $0 + $1.cost.durationSeconds },
                                total: totalLabel
                            )
                        }
                    }

                    Spacer().frame(height: 10)
                    Text(t("usage_footnote"))
                        .wise(Type.tiny)
                        .foregroundStyle(Ink.mute)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.horizontal, 4)
                }
                .padding(.horizontal, 20)
                .padding(.top, 14)
                .padding(.bottom, 24)
            }
        }
        .navigationBarBackButtonHidden()
        .onAppear {
            model.refresh(quiet: true)
            model.loadUsage()
        }
        // Nothing to watch for once nobody is looking at the balance.
        .onDisappear { model.stopWatchingForTopUp() }
    }

    /// The balance, dark so it reads as the headline figure. This is the number
    /// that decides whether the dial button works.
    private var balanceCard: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(t("usage_balance_label")).wise(Type.labelSmall).foregroundStyle(Ink.onDarkMute)
            Spacer().frame(height: 10)
            HStack(alignment: .lastTextBaseline, spacing: 10) {
                Text("\(model.usage.balance)")
                    .wise(Type.display)
                    .foregroundStyle(model.usage.balance == 0 ? Ink.onDarkMute : Ink.onDark)
                Text(t("usage_used_this_month", model.usage.used))
                    .wise(Type.caption)
                    .foregroundStyle(Ink.onDarkMute)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            Spacer().frame(height: 10)
            Text(t(model.usage.balance == 0 ? "usage_balance_empty" : "usage_balance_note"))
                .wise(Type.tiny)
                .foregroundStyle(Ink.onDarkMute)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 16)
        .padding(.vertical, 18)
        .background(Ink.text, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
    }

    /// The web route: it is the only one that works when somebody else is
    /// paying — a parent abroad with no UK card is the case it exists for.
    private var payRow: some View {
        HStack(spacing: 0) {
            VStack(alignment: .leading, spacing: 0) {
                Text(t("usage_other_ways")).wise(Type.rowTitle).foregroundStyle(Ink.text)
                Text(t("usage_other_ways_sub"))
                    .wise(Type.rowSub)
                    .foregroundStyle(Ink.body)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            Spacer().frame(width: 10)
            Text(t("usage_other_copy"))
                .wise(Type.labelSmall)
                .foregroundStyle(Ink.mute)
                .contentShape(Rectangle())
                .onTapGesture {
                    model.topUpLink { url in
                        UIPasteboard.general.string = url
                        copied = true
                        // They may be sending this to somebody else, who could
                        // pay within seconds.
                        model.watchForTopUp()
                    }
                }
            Spacer().frame(width: 16)
            Text(t("usage_other_open"))
                .wise(Type.labelSmall)
                .foregroundStyle(Ink.deep)
                .contentShape(Rectangle())
                .onTapGesture(perform: openPaymentPage)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
    }

    /// Paying happens in the browser and the credits arrive on a webhook, so
    /// there is no result to come back with — the balance is watched instead.
    private func openPaymentPage() {
        model.topUpLink { url in
            guard let link = URL(string: url) else { return }
            UIApplication.shared.open(link)
            model.watchForTopUp()
        }
    }
}

/// One top-up. The price is whatever the server hands back, already formatted —
/// never built here from a number.
private struct PackRow: View {
    let calls: Int
    let price: String?
    let onBuy: () -> Void

    var body: some View {
        HStack {
            Text(t("usage_pack_calls", calls))
                .wise(Type.rowTitle)
                .foregroundStyle(Ink.text)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(price?.nilIfBlank ?? t("usage_pack_price_pending"))
                .wise(Type.labelSmall)
                .foregroundStyle(price?.isNotBlank == true ? Ink.onLime : Ink.mute)
                .padding(.horizontal, 14)
                .padding(.vertical, 8)
                .background(
                    price?.isNotBlank == true ? Ink.lime : Ink.canvasSoft,
                    in: RoundedRectangle(cornerRadius: 10, style: .continuous)
                )
                .contentShape(Rectangle())
                .onTapGesture { if price?.isNotBlank == true { onBuy() } }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
    }
}

private struct UsageRow: View {
    let call: Call
    let symbol: String

    var body: some View {
        HStack(spacing: 0) {
            Text(shortDate(call.createdAt))
                .wise(Type.tiny)
                .foregroundStyle(Ink.mute)
                .frame(width: 38, alignment: .leading)
            Spacer().frame(width: 10)
            Text(call.businessName.isBlank ? call.phoneNumber : call.businessName)
                .wise(Type.listItem)
                .foregroundStyle(Ink.text)
                .lineLimit(1)
                .frame(maxWidth: .infinity, alignment: .leading)
            Spacer().frame(width: 10)
            Text(call.cost.map { durationLabel($0.durationSeconds) } ?? "")
                .wise(Type.mono)
                .foregroundStyle(Ink.mute)
            Spacer().frame(width: 10)
            // Blank rather than £0.00 while Twilio has not rated it: a real zero
            // and a not-yet-known are different things.
            Text(money)
                .wise(Type.mono)
                .foregroundStyle(call.cost != nil ? Ink.text : Ink.rim)
                .frame(width: 46, alignment: .trailing)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }

    private var money: String {
        guard let cost = call.cost else { return t("cost_pending_short") }
        guard let amount = Double(cost.price) else { return "\(symbol)\(cost.price)" }
        return "\(symbol)\(String(format: "%.2f", amount))"
    }

    private func shortDate(_ at: Int) -> String {
        let parts = Calendar.current.dateComponents(
            [.month, .day], from: Date(timeIntervalSince1970: Double(at) / 1000)
        )
        return "\(parts.month ?? 0)/\(parts.day ?? 0)"
    }
}

private struct TotalRow: View {
    let seconds: Int
    let total: String

    var body: some View {
        HStack(spacing: 10) {
            Text(t("usage_total"))
                .wise(Type.value)
                .foregroundStyle(Ink.text)
                .frame(maxWidth: .infinity, alignment: .leading)
            Text(durationLabel(seconds)).wise(Type.mono).foregroundStyle(Ink.mute)
            Text(total).wise(Type.monoBody).foregroundStyle(Ink.text)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 12)
    }
}

File: android/app/src/main/java/com/voicecall/ui/screens/UsageScreen.kt
package com.voicecall.ui.screens

import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import com.voicecall.R
import com.voicecall.data.BillingState
import com.stripe.android.PaymentConfiguration
import com.stripe.android.paymentsheet.PaymentSheet
import com.stripe.android.paymentsheet.PaymentSheetResult
import com.stripe.android.paymentsheet.rememberPaymentSheet
import com.voicecall.data.Call
import com.voicecall.ui.GroupLabel
import com.voicecall.ui.GroupedCard
import com.voicecall.ui.CallsViewModel
import com.voicecall.ui.ScreenHeader
import com.voicecall.ui.clickableNoRipple
import com.voicecall.ui.navBarPadding
import com.voicecall.ui.statusBarPadding
import com.voicecall.ui.theme.Ink
import com.voicecall.ui.theme.Type
import java.util.Calendar

/**
 * What is left to spend, what a top-up costs, and what this month has run to.
 *
 * The money on this screen is deliberately two different things kept apart.
 * The headline is calls — the unit anybody actually buys and spends. Below it,
 * unchanged, is what those calls cost us to run, itemised per call: that is not
 * a bill, and the footnote says so. Showing it is a choice — most services hide
 * their margin — but this app asks people to trust it with a phone line, and a
 * cost breakdown they can check against their own call list is worth more than
 * the margin is worth concealing.
 *
 * Prices for the packs come from Play rather than from here, because Play is
 * the only thing that knows the buyer's currency and country.
 */
@Composable
fun UsageScreen(vm: CallsViewModel, onBack: () -> Unit) {
    val state by vm.uiState.collectAsState()
    val usage by vm.usage.collectAsState()
    val prices by vm.billing.prices.collectAsState()
    val billingState by vm.billing.state.collectAsState()

    // launchBillingFlow wants the Activity itself: Play draws its sheet over
    // this one rather than in it.
    val context = LocalContext.current
    val activity = context as? Activity
    val clipboard = remember(context) {
        context.getSystemService(Context.CLIPBOARD_SERVICE) as? ClipboardManager
    }
    var copied by remember { mutableStateOf(false) }
    val payOnWeb = usage.webPay

    // Stripe's sheet, opened over this screen. Its result is only used to say
    // something on screen — the credits come from Stripe's webhook to our
    // server, never from what the phone reports, because the phone is not who
    // we take payment facts from.
    var payState by remember { mutableStateOf<String?>(null) }
    val paidLabel = stringResource(R.string.usage_pay_done)
    val cancelledLabel = stringResource(R.string.usage_pay_cancelled)
    val paymentSheet = rememberPaymentSheet { result ->
        payState = when (result) {
            is PaymentSheetResult.Completed -> {
                // The webhook is usually a second or two behind the sheet.
                vm.watchForTopUp()
                paidLabel
            }
            is PaymentSheetResult.Canceled -> cancelledLabel
            is PaymentSheetResult.Failed -> result.error.message ?: cancelledLabel
        }
    }

    LaunchedEffect(Unit) {
        vm.refresh(quiet = true)
        // The billing client outlives this screen, so last visit's "added" line
        // would still be sitting there on the next one.
        vm.billing.clearState()
        // Loads the balance and packs, asks Play for prices, and delivers
        // anything paid for on a previous run that never reached the server.
        vm.openShop()
    }
    // Nothing to watch for once nobody is looking at the balance.
    DisposableEffect(Unit) { onDispose { vm.stopWatchingForTopUp() } }

    val monthStart = remember {
        Calendar.getInstance().apply {
            set(Calendar.DAY_OF_MONTH, 1)
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }.timeInMillis
    }
    val thisMonth = state.calls.filter { it.createdAt >= monthStart }.sortedByDescending { it.createdAt }
    val priced = thisMonth.mapNotNull { call -> call.cost?.let { call to it } }
    val total = priced.sumOf { it.second.price.toDoubleOrNull() ?: 0.0 }
    val seconds = priced.sumOf { it.second.durationSeconds }
    // The currency is whatever Twilio rated in, so it is only known once at
    // least one call has been rated. Until then there is no total to show: a
    // bare "0.00" reads as "this month was free" when it means "not priced yet".
    val symbol = priced.firstOrNull()?.second?.unit.orEmpty().let {
        when (it.uppercase()) { "GBP" -> "£"; "USD" -> "$"; "EUR" -> "€"; else -> "" }
    }
    val totalLabel =
        if (priced.isEmpty()) stringResource(R.string.cost_pending_short)
        else "$symbol${"%.2f".format(total)}"

    Column(
        Modifier
            .fillMaxSize()
            .padding(top = statusBarPadding(), bottom = navBarPadding()),
    ) {
        ScreenHeader(title = stringResource(R.string.usage_title), onBack = onBack)

        Column(
            Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState())
                .padding(start = 20.dp, end = 20.dp, top = 14.dp, bottom = 24.dp),
        ) {
            // The balance, dark so it reads as the headline figure. This is the
            // number that decides whether the dial button works.
            Column(
                Modifier
                    .fillMaxWidth()
                    .background(Ink.Text, RoundedCornerShape(16.dp))
                    .padding(horizontal = 16.dp, vertical = 18.dp),
            ) {
                Text(
                    stringResource(R.string.usage_balance_label),
                    style = Type.LabelSmall,
                    color = Ink.OnDarkMute,
                )
                Spacer(Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.Bottom) {
                    Text(
                        "${usage.balance}",
                        style = Type.Display,
                        color = if (usage.balance == 0) Ink.OnDarkMute else Ink.OnDark,
                    )
                    Spacer(Modifier.width(10.dp))
                    Text(
                        stringResource(R.string.usage_used_this_month, usage.used),
                        style = Type.Caption,
                        color = Ink.OnDarkMute,
                        modifier = Modifier.weight(1f).padding(bottom = 4.dp),
                    )
                }
                Spacer(Modifier.height(10.dp))
                Text(
                    stringResource(
                        if (usage.balance == 0) R.string.usage_balance_empty
                        else R.string.usage_balance_note,
                    ),
                    style = Type.Tiny,
                    color = Ink.OnDarkMute,
                )
            }

            // --- the shop ---
            Spacer(Modifier.height(16.dp))
            GroupLabel(stringResource(R.string.usage_shop), modifier = Modifier.padding(bottom = 8.dp))

            // Paid for in-app, through Stripe's own sheet — no browser, no app
            // switch. Shown when the server has it on; the Play packs below
            // take over wherever Play Billing is the configured route.
            if (usage.inAppPacks.isNotEmpty()) {
                GroupedCard(
                    rows = usage.inAppPacks.map { pack ->
                        {
                            PackRow(
                                calls = pack.calls,
                                price = pack.price,
                                enabled = activity != null && payState == null,
                                onBuy = {
                                    payState = context.getString(R.string.usage_paying)
                                    vm.startInAppPayment(pack.priceId) { intent ->
                                        PaymentConfiguration.init(context, intent.publishableKey)
                                        paymentSheet.presentWithPaymentIntent(
                                            intent.clientSecret,
                                            PaymentSheet.Configuration.Builder("Mynah").build(),
                                        )
                                    }
                                },
                            )
                        }
                    },
                )
            }

            if (usage.packs.isEmpty() && usage.inAppPacks.isEmpty()) {
                Text(
                    stringResource(R.string.usage_shop_unavailable),
                    style = Type.Caption,
                    color = Ink.Mute,
                )
            } else if (usage.packs.isNotEmpty()) {
                GroupedCard(
                    rows = usage.packs.map { pack ->
                        {
                            PackRow(
                                calls = pack.calls,
                                // Absent until Play answers. Never a guess: a
                                // price that turns out wrong at the sheet is
                                // worse than one that arrives a moment late.
                                price = prices[pack.productId],
                                enabled = activity != null && billingState !is BillingState.Working,
                                onBuy = { activity?.let { vm.billing.buy(it, pack.productId) } },
                            )
                        }
                    },
                )
            }

            // The web route. Kept alongside the in-app packs rather than hidden
            // behind them: Play cannot offer WeChat Pay or Alipay on a UK
            // account, and for a good few of these users that is the payment
            // method they have. It is also the only route that works when
            // somebody else is paying — see the note under it.
            if (payOnWeb) {
                Spacer(Modifier.height(10.dp))
                GroupedCard(
                    rows = listOf {
                        Row(
                            Modifier
                                .fillMaxWidth()
                                .padding(horizontal = 16.dp, vertical = 14.dp),
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            Column(Modifier.weight(1f)) {
                                Text(
                                    stringResource(R.string.usage_other_ways),
                                    style = Type.RowTitle,
                                    color = Ink.Text,
                                )
                                Text(
                                    stringResource(R.string.usage_other_ways_sub),
                                    style = Type.RowSub,
                                    color = Ink.Body,
                                )
                            }
                            Spacer(Modifier.width(10.dp))
                            Text(
                                stringResource(R.string.usage_other_copy),
                                style = Type.LabelSmall,
                                color = Ink.Mute,
                                modifier = Modifier.clickableNoRipple {
                                    vm.topUpLink { url ->
                                        clipboard?.setPrimaryClip(
                                            ClipData.newPlainText("Mynah top-up", url),
                                        )
                                        copied = true
                                        // They may be sending this to somebody
                                        // else, who could pay within seconds.
                                        vm.watchForTopUp()
                                    }
                                },
                            )
                            Spacer(Modifier.width(16.dp))
                            Text(
                                stringResource(R.string.usage_other_open),
                                style = Type.LabelSmall,
                                color = Ink.Deep,
                                modifier = Modifier.clickableNoRipple {
                                    vm.topUpLink { url ->
                                        // Paying happens in the browser and the
                                        // credits arrive on a webhook, so there
                                        // is no result to come back with — the
                                        // balance is watched for instead.
                                        runCatching {
                                            context.startActivity(
                                                Intent(Intent.ACTION_VIEW, url.toUri()),
                                            )
                                            vm.watchForTopUp()
                                        }
                                    }
                                },
                            )
                        }
                    },
                )
                Spacer(Modifier.height(8.dp))
                Text(
                    stringResource(
                        if (copied) R.string.usage_other_copied else R.string.usage_other_note,
                    ),
                    style = Type.Tiny,
                    color = if (copied) Ink.PositiveDeep else Ink.Mute,
                    modifier = Modifier.padding(horizontal = 4.dp),
                )
            }

            // The in-app sheet has its own word to say; the Play flow's notice
            // follows underneath it.
            val notice = payState?.let { it to Ink.Body } ?: when (val s = billingState) {
                is BillingState.Working -> stringResource(R.string.usage_working) to Ink.Body
                is BillingState.Bought ->
                    stringResource(R.string.usage_bought, usage.balance) to Ink.PositiveDeep
                is BillingState.Failed -> s.reason to Ink.NegativeDeep
                BillingState.Idle -> null
            }
            notice?.let { (message, colour) ->
                Spacer(Modifier.height(10.dp))
                Text(message, style = Type.Caption, color = colour, modifier = Modifier.padding(horizontal = 4.dp))
            }

            // --- what the month cost to run ---
            Spacer(Modifier.height(22.dp))
            GroupLabel(stringResource(R.string.usage_breakdown), modifier = Modifier.padding(bottom = 8.dp))

            if (thisMonth.isEmpty()) {
                Text(stringResource(R.string.usage_empty), style = Type.Caption, color = Ink.Mute)
            } else {
                val rows = buildList<@Composable () -> Unit> {
                    thisMonth.forEach { call -> add { UsageRow(call, symbol) } }
                    add { TotalRow(seconds, totalLabel) }
                }
                GroupedCard(rows = rows)
            }

            Spacer(Modifier.height(10.dp))
            Text(
                stringResource(R.string.usage_footnote),
                style = Type.Tiny,
                color = Ink.Mute,
                modifier = Modifier.padding(horizontal = 4.dp),
            )
        }
    }
}

/**
 * One top-up. The price is whatever Play hands back, already formatted in the
 * buyer's own currency — never built here from a number.
 */
@Composable
private fun PackRow(calls: Int, price: String?, enabled: Boolean, onBuy: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 14.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            stringResource(R.string.usage_pack_calls, calls),
            style = Type.RowTitle,
            color = Ink.Text,
            modifier = Modifier.weight(1f),
        )
        Box(
            Modifier
                .background(
                    if (enabled && price != null) Ink.Lime else Ink.CanvasSoft,
                    RoundedCornerShape(10.dp),
                )
                .then(if (enabled && price != null) Modifier.clickableNoRipple(onBuy) else Modifier)
                .padding(horizontal = 14.dp, vertical = 8.dp),
        ) {
            Text(
                price ?: stringResource(R.string.usage_pack_price_pending),
                style = Type.LabelSmall,
                color = if (enabled && price != null) Ink.OnLime else Ink.Mute,
            )
        }
    }
}

@Composable
private fun UsageRow(call: Call, symbol: String) {
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(shortDate(call.createdAt), style = Type.Tiny, color = Ink.Mute, modifier = Modifier.width(38.dp))
        Spacer(Modifier.width(10.dp))
        Text(
            call.businessName.ifBlank { call.phoneNumber },
            style = Type.ListItem,
            color = Ink.Text,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f),
        )
        Spacer(Modifier.width(10.dp))
        Text(
            call.cost?.let { duration(it.durationSeconds) }.orEmpty(),
            style = Type.Mono,
            color = Ink.Mute,
        )
        Spacer(Modifier.width(10.dp))
        Text(
            // Blank rather than £0.00 while Twilio has not rated it: a real
            // zero and a not-yet-known are different things.
            call.cost?.let { "$symbol${it.price.toDoubleOrNull()?.let { p -> "%.2f".format(p) } ?: it.price}" }
                ?: stringResource(R.string.cost_pending_short),
            style = Type.Mono,
            color = if (call.cost != null) Ink.Text else Ink.Rim,
            modifier = Modifier.width(46.dp),
        )
    }
}

@Composable
private fun TotalRow(seconds: Int, totalLabel: String) {
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(stringResource(R.string.usage_total), style = Type.Value, color = Ink.Text, modifier = Modifier.weight(1f))
        Text(duration(seconds), style = Type.Mono, color = Ink.Mute)
        Spacer(Modifier.width(10.dp))
        Text(totalLabel, style = Type.MonoBody, color = Ink.Text)
    }
}

private fun duration(seconds: Int): String =
    if (seconds >= 60) "%dm %02ds".format(seconds / 60, seconds % 60) else "${seconds}s"

private fun shortDate(at: Long): String = Calendar.getInstance().apply { timeInMillis = at }.let {
    "%d/%d".format(it.get(Calendar.MONTH) + 1, it.get(Calendar.DAY_OF_MONTH))
}
